### DISCUSSION TITLE
Redirection with proxy

### DISCUSSION THREAD

**Discussion by MinerPL** [nodejs/undici#1925] [Q&A] [answered]
When trying to make a request to a page with a redirect (e.g. http -> https) undici returns me an error 

```
TypeError: fetch failed
    at fetch (C:\Users\minerpl\IdeaProjects\MightyBot\node_modules\undici\index.js:113:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async C:\Users\minerpl\IdeaProjects\MightyBot\test-proxy.js:25:17 {
  cause: Error: Request was cancelled.
```

If I make a direct link from https everything works as expected. Do I need to add something to the fetch option to make it work?

**Comment by metcoder95** [comment]
While using fetch, in case of your request should expect any redirect, you need to specify to `fetch` what to do. in this case, you can either indicate to `follow` the redirect, `error`, or `manual`ly follow the redirection.

See: https://developer.mozilla.org/en-US/docs/Web/API/fetch

**Reply by MinerPL** [reply]
The follow option doesn't work when trying to use it with ProxyAgent. It returns an error as above.

**Reply by metcoder95** [reply]
Could you provide a 
[Minimum Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example) to help you in a better way?

**Reply by MinerPL** [reply]
Yes, the code it tries to execute is below.

```js
    const dispatcher = new ProxyAgent({
        uri: 'http://proxyip:proxyport/',
        auth: Buffer.from(`proxylogin:proxypassword`).toString('base64'),
    });

    const fetchOptions = {
        redirect: 'follow',
        headers: {
            'User-Agent': "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.109 Safari/537.36",
            'Accept': 'text/html',
        },
        dispatcher: dispatcher instanceof ProxyAgent ? dispatcher : null,
    }

    const res = await fetch('http://google.com', fetchOptions)
        .then((res) => res.text())
        .then(console.log)
        .catch(console.error);
```

**Reply by metcoder95** [reply]
Can you try with `Client` or `request` instead?
I've tried to reproduce it and found it only when the Proxy server was not behaving as expected.

Although I have to admit that the `fetch` error is not really meaningful 🙁

**Reply by MinerPL** [reply]
Using `request` inside of `fetch` undici throw error about response status
```
RequestAbortedError [AbortError]: Proxy response !== 200 when HTTP Tunneling
    at Client.connect (PROJECT\node_modules\undici\lib\proxy-agent.js:93:22)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 'UND_ERR_ABORTED'
}
```

Could the problem be that the proxies are http and not https?

**Reply by metcoder95** [reply, marked as answer]
Not really sure, I'm not aware of the settings for the Proxy server, it can be your proxy is not behind HTTPs, or even that the credentials are wrong. I'll suggest you look into that direction 🙂

**Reply by MinerPL** [reply]
The credentials are correct (requests that do not have a redirect work without a problem), the https itself works only the problem arises when redirecting from http to https. I'll try running a proxy on https and see if that solves the redirect problem.

**Reply by MinerPL** [reply]
After checking the proxy configuration it turned out that the proxy was the problem, the proxy server is a squid and in the default configuration it has this configuration option

```
# Deny CONNECT to other than secure SSL ports
http_access deny CONNECT !SSL_ports
```

After commenting it out, fetch works correctly with redirects.
